### PR TITLE
Some Position utility methods for approximate offsets

### PR DESCRIPTION
**Pull request by daboross**
Just some methods which could be useful?

I'm on the edge for whether these do belong in this library, but if they're useful to people they shouldn't hurt _too much_.

If not, maybe now would be a good time to add a "position_utils" addon library?

### CODE REVIEW

File: src/local/room_position/approximate_offsets.rs [hunk 1]
```diff
@@ -0,0 +1,147 @@
+//! Methods related to approximating positions between other positions.
+use crate::objects::HasPosition;
+
+use super::Position;
+
+impl Position {
+    /// Calculates an approximate midpoint between this point and the target.
```
Comment (ASalvail) [inline src/local/room_position/approximate_offsets.rs]: This is not a midpoint (as in, it's not in the middle). It's `distance_towards_target` from this point towards the other point. (with big caveats)

Comment (ASalvail) [reply src/local/room_position/approximate_offsets.rs]: Also, this line is suspiciously like the first one of the next function 😛

File: src/local/room_position/approximate_offsets.rs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        self + (new_offset_x, new_offset_y)
+    }
+
+    /// Calculates an approximate midpoint between this point and the target.
+    ///
+    /// In case of a tie, rounds towards the target.
+    ///
+    /// If `distance_from_target` is bigger than the distance to the target,
+    /// this position is returned.
+    pub fn between<T>(self, target: &T, distance_from_target: i32) -> Position
```
Comment (ASalvail) [inline src/local/room_position/approximate_offsets.rs]: Given that the other is `toward`, wouldn't this be more understandable as `from`?

File: src/local/room_position/approximate_offsets.rs [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    where
+        T: ?Sized + HasPosition,
+    {
+        target.pos().towards(&self, distance_from_target)
+    }
+
+    /// Calculates an approximate midpoint between this point and the target.
+    ///
+    /// In case of a tie, rounds towards the target.
+    pub fn midpoint_between<T>(self, target: &T) -> Position
```
Comment (ASalvail) [inline src/local/room_position/approximate_offsets.rs]: suggestion: `midpoint_between` -> `midpoint_to`
It reads better.

File: src/local/room_position/approximate_offsets.rs (hunk 1 above)
Comment (ASalvail) [reply src/local/room_position/approximate_offsets.rs]: Maybe: `midpoint` -> `offset`

File: src/local/room_position/approximate_offsets.rs [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            return target;
+        }
+
+        let new_offset_x = (offset_x * distance_towards_target) / total_distance;
+        let new_offset_y = (offset_y * distance_towards_target) / total_distance;
+
+        self + (new_offset_x, new_offset_y)
+    }
+
+    /// Calculates an approximate midpoint between this point and the target.
```
Comment (ASalvail) [inline src/local/room_position/approximate_offsets.rs]: Same: ~midpoint~ offset

File: src/local/room_position/approximate_offsets.rs [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    use super::Position;
+    use crate::RoomName;
+
+    fn test_rooms() -> impl Iterator<Item = RoomName> {
+        ["E0N0", "E20N20", "W20N0", "E20S20", "W20S20"]
+            .iter()
+            .map(|s| s.parse().unwrap())
+    }
+
+    fn pos(room: RoomName, x: u32, y: u32) -> Position {
```
Comment (ASalvail) [inline src/local/room_position/approximate_offsets.rs]: `use super::Position::new as pos;`? 😛

**Comment by daboross** [comment]
Sorry I've just kind of left this. Really glad to have your review, and once I have the time to be more active again, I'll be fixing the things! (and possibly just sending the code into a separate crate anyways...)

**Review by ASalvail — APPROVED** [review]

**Comment by ASalvail** [comment]
Just for cleanup's sake.

**Pull request merged**